// markdown-table-prettify/cli/argumentsParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseArguments = void 0;
function parseArguments(processArgs) {
    return {
        check: Boolean(hasArgument(ArgumentNames.CHECK_ARG) || false),
        columnPadding: Number(getArgumentValue(ArgumentNames.PADDING_ARG) || 0)
    };
    function hasArgument(key) {
        return processArgs.length > 2 && processArgs.find(function (arg) { return arg.startsWith("--" + key); }) !== undefined;
    }
    function getArgumentValue(key) {
        var hasArguments = processArgs.length > 2;
        var split = (hasArguments
            ? processArgs.find(function (arg) { return arg.startsWith("--" + key); }) || ""
            : "")
            .split("=");
        return split.length == 2
            ? split[1]
            : null;
    }
}
exports.parseArguments = parseArguments;
var ArgumentNames = /** @class */ (function () {
    function ArgumentNames() {
    }
    ArgumentNames.CHECK_ARG = "check";
    ArgumentNames.PADDING_ARG = "columnPadding";
    return ArgumentNames;
}());
